File: Cpp/AdventOfCode/src/2015/d10_SeeAndSay.cpp
#pragma once
#include "2015/d10_SeeAndSay.h"

SOLUTION(2015, 10) {
    constexpr void AppendNext(char lastCharacter, int count, std::string & result) {
        switch (count) {
        case 1: result += "1"; break;
        case 2: result += "2"; break;
        case 3: result += "3"; break;
        }
        result += lastCharacter;
    }

    constexpr std::string Next(const std::string & input) {
        std::string result;
        char lastCharacter = input[0];
        int count = 0;
        for (const auto& c : input) {
            if (c == lastCharacter) {
                count++;
            }
            else {
                AppendNext(lastCharacter, count, result);
                lastCharacter = c;
                count = 1;
            }
        }
        AppendNext(lastCharacter, count, result);

        return result;
    }

    constexpr std::string ApplyTimes(std::string_view input, size_t times) {
        std::string running = std::string(input);

        for (auto i = 0; i < times; i++) {
            running = Next(running);
        }

        return running;
    }


    PART_ONE() {
        return Constexpr::ToString(ApplyTimes(Line, 40).size());
    }
    PART_TWO() {
        return Constexpr::ToString(ApplyTimes(Line, 50).size());
    }
    TESTS() {
        static_assert(Next("1") == "11");
        static_assert(Next("11") == "21");
        static_assert(Next("21") == "1211");
        static_assert(Next("1211") == "111221");
        static_assert(Next("111221") == "312211");
        //static_assert(ApplyTimes("1113222113", 40).size() == 252594);

        return true;
    }
}

File: Rust/rust_aoc/src/year_2015/d10_see_and_say.rs
pub fn part_one(lines: &Vec<String>) -> String {
    apply_times(&lines[0], 40).len().to_string()
}

pub fn part_two(lines: &Vec<String>) -> String {
    apply_times(&lines[0], 50).len().to_string()
}

fn apply_times(start: &str, times: u32) -> String {
    let mut running = String::from(start);
    for _i in 0..times {
        running = next(&running);
    }
    running
}

fn next(input: &str) -> String {
    let mut result: String = String::new();
    let mut last_char = input.chars().next().unwrap();
    let mut count = 0u32;
    let mut buf = [0; 2];

    for c in input.chars() {
        if c == last_char {
            count += 1;
        } else {
            let to_add = count.to_string() + last_char.encode_utf8(&mut buf);
            result += &to_add;
            last_char = c;
            count = 1;
        }
    }

    let to_add = count.to_string() + last_char.encode_utf8(&mut buf);
    result += &to_add;

    result
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn next_tests() {
        assert_eq!(next("1"), "11");
        assert_eq!(next("11"), "21");
        assert_eq!(next("21"), "1211");
        assert_eq!(next("1211"), "111221");
        assert_eq!(next("111221"), "312211");
    }
}